Fix: Shadow/Font size preset panel crashes the editor

t-hamano opened the pull request:
Temporary fix for #65711
Related to #64777

## What?
This PR fixes a critical error that occurs when returning to the previous screen from the shadow presets panel or the font presets panel.

This fix isn't ideal, but we need to prevent the issue from shipping in the upcoming Gutenberg 19.4.

## Why?

I found that the problem was caused by https://github.com/WordPress/gutenberg/pull/64777.

From [this comment](https://github.com/WordPress/gutenberg/issues/65711#issuecomment-2380625635):

> The cause of this problem is likely that in the component where the crash occurs, some logic relies on URL parameters, namely `useNavigator`.
> 
> For example, in the Shadows panel, `category` and `slug` are taken from URL parameters [here](https://github.com/WordPress/gutenberg/blob/faa6968936e8d996acafb8fd846d3ca08dad7088/packages/edit-site/src/components/global-styles/shadows-edit-panel.js#L75-L78).
> 
> However, when we return from this panel, the URL parameters have already been changed to new ones, so all parameters are `undefined`. The panel crashes because parameters are not expected to be `undefined`.

## How?

Perform an early return to avoid a critical error when the parameter is undefined. One ideal future solution is suggested in [this comment](https://github.com/WordPress/gutenberg/issues/65711#issuecomment-2383603543).

## Testing Instructions

- Open the site editor > global styles > Shadows > Choose one of the presets > click the Back button
- Open the site editor > global styles > Typography > Font size presets > Choose one of the presets > click the Back button

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/013063ee-49f5-420c-8405-5627e26f3709

t-hamano committed:
Fix: Shadow/Font size preset panel crashes the editor

tyxla reviewed (approved):
This looks good 👍 Thanks for the fix!

Long-term, I agree with @ciampo that ideally the component would handle this - errors like these are consequences of keeping an open API with regards to those params.

ciampo reviewed (approved):
Agreed. LGTM as an urgent fix. I'll see if I can create a slightly more sophisticated fix , still in app-land — it will help us to understand how to handle this better at the `Navigator`-level.

The E2E failure doesn't seem related, and can be encountered across many other PRs.

t-hamano commented:
Thanks for the review!

I think the E2E test failure is related to the core, so I would like to merge it once the core is fixed.

Core ticket: https://core.trac.wordpress.org/ticket/60666

t-hamano commented:
Another critical error was found regarding font size presets: https://github.com/WordPress/gutenberg/issues/65774

I believe that the issue is not related to #64777, but simply a problem with the font size preset logic itself, and it also occurs in WP 6.7. On the other hand, this PR does not need to be backported to WP6.7.

Therefore, it might be better to merge this PR _after_ backporting the fix for #65774 to WP 6.7.

ramonjd commented on packages/edit-site/src/components/global-styles/font-sizes/font-size.js:
```diff
@@ -46,6 +46,10 @@ function FontSize() {
 
 	const [ globalFluid ] = useGlobalSetting( 'typography.fluid' );
 
+	if ( ! origin || ! slug ) {
```
Oh nice! I promise I didn't steal this for https://github.com/WordPress/gutenberg/pull/65811

😆

ramonjd reviewed (approved):
Tested quickly in playground. Thanks for catching this so quickly.

ciampo commented:
> Therefore, it might be better to merge this PR _after_ backporting the fix for #65774 to WP 6.7.

Should we rebase, re-test and then merge ?

t-hamano committed:
Merge branch 'trunk' into fix/shadow-font-size-crash

t-hamano committed:
Refactor logic

t-hamano commented:
I updated this PR as#65791 was merged.

The following warning was raised, so I moved the location of `useEffect` and merged the two early returns:

```
React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return?
```

 The issue reported in #65711 has been resolved as before in this PR, and #65774 should not have reoccurred:

https://github.com/user-attachments/assets/ccbf5921-5abf-41cf-bef8-e3176dcb5371

ciampo reviewed (approved):
LGTM 🚀 

As mentioned before, the real fix should be applied in `Navigator`. Once that's done, we'll be able to rework the code, removing the early returns (and using `goTo` only as a fail-safe for other use-cases)
